### PR TITLE
sdk/trace: removing ApplyConfig and Config

### LINKED ISSUE #1636: Remove TracerProvider.ApplyConfig and Config
The `TracerProvider.ApplyConfig` method and related `Config` are additions to the exposed API that likely are not needed nor ideal. The `ApplyConfig` method exists to update a `TracerProvider`'s configuration. This is an optional part of [the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.0.1/specification/trace/sdk.md#tracer-creation):

> The TracerProvider MAY provide methods to update the configuration.

meaning that we could remove these items and remain compliant with the specification.

Removing this type and method from the exported API will be a reduction in the API. This will reduce the on-boarding load for new users wondering when they should be configuring the `TracerProvider` (at creation or after with this method). Additionally, it could reduce the maintenance burden if we decide in the future to add direct Setters for the `Config` fields to avoid issues similar to the ones identified in #1631 but would still need to support this exported method and type.

Removing these will remove functionality from the `TracerProvider` API, the user will no longer be able to update a configuration of a TracerProvider after it has been created. There do not seem to be any reasons why we would need to update this configuration after creation so this seems justified. If this functionality needs to be added back in we could likely do so in a way that better serves the intended purpose and one that doesn't leak internal implementation details of how the `TracerProvider` manages a synchronized configuration.

### Proposal

Removing the `ApplyConfig` method from the `TracerProvider` will make the configuration of span limits, the default sampler, ID generators, and resource immutable. This means that the `config` field of the `TracerProvider` which holds an atomic value of the related `Config` could be flattened into the fields of a `Config` directly.

Removing this field and with the merging of #1633 the `Config` type will be unused. It can be removed when this is done.

### PR DESCRIPTION
**Pull request by ijsong**
This patch removes `ApplyConfig` method and `Config` struct from `go.opentelemetry.io/otel/sdk/trace` package.  To ensure valid config for TracerProvider, it adds `ensureValidTracerProviderConfig` private function.

Jaeger and Zipkin have been used the `Config` directly across package boundaries. Since `Config` is removed, they can't use it. This change, thus, replaces `WithSDK` with `WithSDKOptions`.

Resolves #1636, #1705.

### CODE REVIEW

File: exporters/trace/jaeger/jaeger.go
Comment (Aneurysm9) [inline exporters/trace/jaeger/jaeger.go:63]: Replicating the `sdktrace.TracerProviderOption`s here makes me somewhat uneasy.  Following this pattern would require that they be duplicated in every exporter and that any new `TracerProviderOption`s be added in multiple places.  I wonder if instead a `WithSDKOptions(opts ...sdktrace.TracerProviderOption)` would be better.  Then we could store a `[]sdktrace.TracerProviderOption` in the exporter config and append `sdktrace.WithSyncer(exporter)` to it just before calling `sdktrace.NewTracerProvider(pOpts...)`

File: exporters/trace/zipkin/model.go
Comment (Aneurysm9) [inline exporters/trace/zipkin/model.go:153]: Should this happen before adding values from `data.Attributes`?  I think we expect that `Resource` values will underlay user-added `Attribute` values.

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
I think I'd like to see a different solution to getting `TracerProviderOption`s from the exporter pipeline convenience methods to the `TracerProvider` initialization before landing this.

File: exporters/trace/jaeger/jaeger.go
Comment (ijsong) [reply exporters/trace/jaeger/jaeger.go]: Done.

File: exporters/trace/zipkin/model.go
Comment (ijsong) [reply exporters/trace/zipkin/model.go]: Done.

**Review by Aneurysm9 — APPROVED** [review]

**Comment by Aneurysm9** [comment]
@ijsong sorry about the churn on this, there were several other issues working in the same area though I think we've made it through most of them.

@open-telemetry/go-approvers can we get another approval on this so we can land it?

**Comment by ijsong** [comment]
@Aneurysm9 No problem, I fixed some issues after merging main branch, and renamed field name `defaultSampler` to `sampler` to prepare #1702.

File: exporters/trace/zipkin/model.go
Comment (ijsong) [reply exporters/trace/zipkin/model.go]: I reverted this change, since adding `Resource` to zipkin tags is not mentioned in the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk_exporters/zipkin.md#summary).

> TBD : This is work in progress document and it is currently doesn't specify mapping for these fields:
> 
> OpenTelemetry fields:
> 
> Resource attributes

**Review by XSAM — APPROVED** [review]

File: exporters/trace/zipkin/zipkin_test.go
Comment (MrAlias) [inline exporters/trace/zipkin/zipkin_test.go:415]:
```suggestion
	require.NoError(t, tp.Shutdown(context.Background))
```

File: exporters/trace/zipkin/zipkin_test.go
Comment (MrAlias) [inline exporters/trace/zipkin/zipkin_test.go:421]: The shutdown (or force flush) above should flush the exporter and not require this wait here. If this does indeed require a wait it will introduce a bug on slow CI systems. Can we please verify this is not needed and remove it? Or replace it with a deterministic signal of some sort?

File: sdk/trace/provider.go
Comment (MrAlias) [inline sdk/trace/provider.go:45]: This comment just reiterates the fact this field is unexported.

```suggestion
	// idGenerator is used to generate all Span and Trace IDs when needed.
```

File: sdk/trace/provider.go
Comment (MrAlias) [inline sdk/trace/provider.go:48]:
```suggestion
	// spanLimits defines the attribute, event, and link limits for spans.
```

File: sdk/trace/provider.go
Comment (MrAlias) [inline sdk/trace/provider.go:307]: Move validation of span limits to its own function.

File: sdk/trace/span.go
Comment (MrAlias) [inline sdk/trace/span.go:610]: Remove makeSamplingDecision and duplicate samplingData struct.

This has become a convoluted function and struct that duplicates `SamplingParameters` and embeds the singular method the function calls within the struct. It does not appear necessary anymore.

File: sdk/trace/provider.go
Comment (MrAlias) [reply sdk/trace/provider.go:307]: https://github.com/open-telemetry/opentelemetry-go/issues/1705

File: sdk/trace/span.go
Comment (MrAlias) [reply sdk/trace/span.go:610]: https://github.com/open-telemetry/opentelemetry-go/issues/1706

File: exporters/trace/jaeger/jaeger_test.go
Comment (ijsong) [reply exporters/trace/jaeger/jaeger_test.go:898]: It should be a pointer of the slice since slice `batchesUploaded` is expanded by `(*testCollectorEnpointWithBatches).upload` method to read later. 
But to simplify code, I am changing to reuse already exist struct - testCollectorEnpoint.

File: exporters/trace/zipkin/zipkin_test.go
Comment (ijsong) [reply exporters/trace/zipkin/zipkin_test.go:420]: My comment is incorrect. The BatshSpanProcessor exports the span correctly when shutting down. Only need time for `mockZipkinCollector` to receive it. 
I change the test time to a second as other tests.

File: sdk/trace/span.go
Comment (ijsong) [reply sdk/trace/span.go:610]: It'd be better to separate PR. Removing `makeSamplingDecision` will result in removing struct `samplingData` because it has no use.

File: sdk/trace/provider.go
Comment (ijsong) [reply sdk/trace/provider.go]: Done.

File: sdk/trace/span.go
Comment (MrAlias) [reply sdk/trace/span.go:610]: > It'd be better to separate PR.

Sorry, that wasn't clear :facepalm:, that was what I was hoping as well.
 
> Removing `makeSamplingDecision` will result in removing struct `samplingData` because it has no use.

Agreed, that would be ideal :+1: 

To be clear we should leave this for a follow on PR that resolves #1706

File: exporters/trace/zipkin/zipkin_test.go
Comment (MrAlias) [inline exporters/trace/zipkin/zipkin_test.go:419]: `tp.Shutdown` should flush all spans, is this needed?

File: sdk/trace/provider.go
Comment (MrAlias) [reply sdk/trace/provider.go]: Sorry, meant for this to be tackled in separate PR.

File: exporters/trace/zipkin/zipkin_test.go
Comment (ijsong) [reply exporters/trace/zipkin/zipkin_test.go:419]: Yes, to export span as soon as possible, `tp.Shutdown` is an easy way to do, I think.
However, if this test uses Exporter directly rather than `NewExportPipeline`, it can use `ExportSpans`. If using `tp.Shutdown` is not good, I will change it to use Exporter.

Comment (MrAlias) [reply exporters/trace/zipkin/zipkin_test.go:419]: I'm confused, TracerProvider.Shutdown will call the SpanProcessor Shutdown which calls the exporter Shutdown.

Have you tried removing this and it failed to receive spans?

Comment (ijsong) [reply exporters/trace/zipkin/zipkin_test.go:419]: oops, sorry for my mistake. I confused with `tp.ForceFlush`. I will remove `require.Eventually(t, checkFunc, time.Second, 10*time.Millisecond)`.

**Review by MrAlias — APPROVED** [review]

File: exporters/trace/zipkin/zipkin_test.go
Comment (ijsong) [reply exporters/trace/zipkin/zipkin_test.go]: Done.

**Pull request merged**
